### ISSUE TITLE
Provision an AWS cluster for Thoth-station

### ISSUE THREAD

**Issue by harshad16** [operate-first/support#396]
**Is your feature request related to a problem? Please describe.**
The thoth-station production environment has been down for nearly 2 weeks now (after we lost zero cluster), we would like to have an AWS cluster provisioned as the Smaug cluster seems to be in issues. 
This would allow us to serve our users. 

Requirements:

Openshift 4.7 or above

Minimum Cluster requirement:
CPU: 72 
Memory: 128 Gi
Storage: 120 Gi

S3 storage: amazon 

Please help us with the provision and management of the cluster.

**Comment by HumairAK** [comment]
Looking at the ec2 instance types, this seems to fit for the requirement: 

For master nodes (3 nodes): 
m5.2xlarge - 8vCPU, 32GiB RAM 

For workers (3 nodes)
c5.9xlarge - 36vCPU, 72GiB RAM - Compute Optimized 

300 GB storage per node 

For context, we currently have a cluster that's: 
3 x m5.2xlarge + 3 x m5.8xlarge (100 GB SSD each node) 

After running for about a day, the costs for this were approximately $120/day


@durandom @goern  -- we can spin this up for a long running cluster, wdyt? (open to suggestions for other instance types)

**Comment by harshad16** [comment]
@codificat  yes you are right, i have updated the description now :+1: 
thanks

**Comment by HumairAK** [comment]
Haven't tried configuring aws s3, but maybe we can get more info? 
How much s3 storage is needed?

**Comment by harshad16** [comment]
Is AWS s3 a different configuration? 
maybe we can start with 50Gb and see about it from there.

**Comment by codificat** [comment]
I don't think it's about (pre-)provisioning... it would be about using the AWS S3 API endpoint instead of Ceph (meaning you don't have to install Ceph). Makes sense?

**Comment by goern** [comment]
Please go ahead.

In any case we are happy to move the workload to a more cost efficient cluster when available...

**Comment by codificat** [comment]
One suggestion: when provisioning the cluster, could you add the custom tag `openshiftClusterID` to the installer, with a value of your choice (e.g. `thoth-station`)? You can specify this in [`platform.aws.userTags`](https://docs.openshift.com/container-platform/4.8/installing/installing_aws/installing-aws-customizations.html#installation-configuration-parameters-optional-aws_installing-aws-customizations) in the installer configuration.

This is for cost tracking purposes: we are trying that tag on another account; it's still not settled, but it might be helpful soon - and it should hopefully be easy and harmless.

**Comment by durandom** [comment]
Can I suggest to use / resize the workshop cluster for this? 
The we have only one cluster, unless there are technical reasons for slicing by cluster

**Comment by tumido** [comment]
@HumairAK I'd prefer more of a smaller worker nodes just because of the cost efficiency when auto scaling the cluster - also let's rather provision a cluster that is smaller than the initial demand and autoscale it based of the actual usage.

The proposed worker nodes cost ~$40 a day that is much more expensive to autoscale than something that cost ~$10... what about `c5.2xlarge` or `c5.4xlarge` instead?

@durandom I think it may be better to start fresh with different instance types - more CPU focused nodes in this case (compared to the balanced we use for workshops) since we don't want to be paying for too much memory sitting around idle.

Also -  @HumairAK the storage per node has no relation to the usable storage by users - this on-node storage serves for local image registry, node binaries and what not. User storage is provisioned via EBS, so there's no real cap on it and we pay for that extra on top of the compute.

**Comment by HumairAK** [comment]
> @HumairAK I'd prefer more of a smaller worker nodes just because of the cost efficiency when auto scaling the cluster

@tumido yeah initially thought this too. Not against it -- but their largest CPU or Memory heavy pod should be able to fit on a node.

```
c5.2xlarge: vCPU 8, MEM 16GiB
c5.4xlarge: vCPU 16, MEM 32GiB

```
@harshad16 any pods using more than 8 cores and 16GiB? 

> the storage per node has no relation to the usable storage by users - this on-node storage serves for local image registry, node binaries and what not. User storage is provisioned via EBS, so there's no real cap on it and we pay for that extra on top of the compute.

Good to know, thanks!

**Comment by harshad16** [comment]
@HumairAK we have one pod which touches about 6cpu and 12 Gb memory.

**Comment by HumairAK** [comment]
Sounds like `c5.2xlarge` should be sufficient then.

**Comment by 4n4nd** [comment]
I think we should do 2 worker nodes of `c5.9xlarge - 36vCPU, 72GiB RAM - Compute Optimized`.
1. This will provide the users (thoth) with more flexibility for larger pod scheduling
2. 2 nodes will satisfy the minimum cpu requirement (i.e. 72 cpus). 
3. If needed more in the future, we can add smaller nodes 

If we do `c5.2xlarge: vCPU 8, MEM 16GiB`, we will need 9 nodes to satisfy the minimum requirements.
1. We will be paying for storage on each node
2. If the users have a bunch of medium sized pods (using 2-3 cpus or even 4-5 GiB memory), this will limit the nodes where the large pods can be deployed on

**Comment by tumido** [comment]
> I think we should do 2 worker nodes of `c5.9xlarge - 36vCPU, 72GiB RAM - Compute Optimized`.

Let's not go with less than 3 nodes as the bare minimum due to availability zones. Each region has 3 avail zones, so let's have at least a node in each.

> If we do `c5.2xlarge: vCPU 8, MEM 16GiB`, we will need 9 nodes to satisfy the minimum requirements.

Is it really a bare minimum or peak usage though? I think @Harshad's request means "this is the bare minimum to handle our peak usage", not that this is consumed at every time. So I expect that we'll need 3-4 nodes in off-peak times and 6-9 at peak... Which is fine.

I think `c5.4xlarge` can be a good middle ground then, wdyt?
 
> 1. We will be paying for storage on each node

True

> 2. If the users have a bunch of medium sized pods (using 2-3 cpus or even 4-5 GiB memory), this will limit the nodes where the large pods can be deployed on

Kube scheduler can handle it and condense smaller pods to a single node I think.

**Comment by 4n4nd** [comment]
> I think `c5.4xlarge` can be a good middle ground then, wdyt?

Yeah that sounds good to me :+1:

**Comment by harshad16** [comment]
Sounds good to me, when can we expect this to be available ?

**Comment by HumairAK** [comment]
So to summarize, we're looking at: 
3 x m5.2xlarge (masters)
5 x c5.4xlarge  (workers)

Correct?

**Comment by harshad16** [comment]
sounds good to me :+1:

**Comment by tumido** [comment]
Yes :+1:  :slightly_smiling_face: Let's o it :slightly_smiling_face:

**Comment by goern** [comment]
cool, who will do it?

**Comment by durandom** [comment]
Is the process documented? If so maybe somebody who hasn't done it yet, could run with it and improve it.

**Comment by HumairAK** [comment]
what should the name of this cluster be?

**Comment by tumido** [comment]
Moria?

**Comment by harshad16** [comment]
> what should the name of this cluster be?
> 
> how about another LOTR name? Elrond? Gimli? :)

**Comment by goern** [comment]
Ozora?! like in https://www.youtube.com/watch?v=123Z5KLLiO8 ?

**Comment by HumairAK** [comment]
That's not Lord of the Rings :roll_eyes:

**Comment by harshad16** [comment]
We would like to go with **balrog**
@HumairAK please let us know what is the step in the procedure.

**Comment by HumairAK** [comment]
the cluster is being provisioned

**Comment by harshad16** [comment]
Thanks @HumairAK

**Comment by HumairAK** [comment]
> Is the process documented? If so maybe somebody who hasn't done it yet, could run with it and improve it.
@durandom 
https://github.com/operate-first/support/pull/399

**Comment by HumairAK** [comment]
> One suggestion: when provisioning the cluster, could you add the custom tag openshiftClusterID to the installer, with a value of your choice (e.g. thoth-station)? You can specify this in platform.aws.userTags in the installer configuration. 

bah I missed this, can this be added post install?

**Comment by 4n4nd** [comment]
> > One suggestion: when provisioning the cluster, could you add the custom tag openshiftClusterID to the installer, with a value of your choice (e.g. thoth-station)? You can specify this in platform.aws.userTags in the installer configuration.
> 
> bah I missed this, can this be added post install?

Added tag `openshiftClusterID=thoth-station` to all the balrog EC2 Instances, Volumes and Elastic IPs.
I don't think this tag will be added automatically for any volumes that are provisioned later for PVCs. 
@codificat is there any way to do that?

**Comment by codificat** [comment]
> Added tag `openshiftClusterID=thoth-station` to all the balrog EC2 Instances, Volumes and Elastic IPs.

Thanks!

> I don't think this tag will be added automatically for any volumes that are provisioned later for PVCs.
> @codificat is there any way to do that?

I'm not aware of a way to do this automagically outside of the installer.

That's fine though. This should already cover the highest costs.

More info: the tag is not yet part of the established reporting, it's a "PoC" that is being tried by a sister AWS account. Having it in these resources opens the door to include it in that experiment and hopefully provide more details on the associated costs for this cluster.

**Comment by HumairAK** [comment]
Okay I created an iam user/bucket policy/bucket and sent @harshad16 the access/secret key and confirmed it works. 

I think that about does it for this cluster right? Can we close this?
